### PR TITLE
Overload prevention port from Java

### PR DESCRIPTION
**Pull request by stefaneberl**

### CODE REVIEW

File: src/openkit-shared/Core/Communication/BeaconSendingCaptureOnState.cs [hunk 1]
```diff
@@ -63,28 +67,61 @@ protected override void DoExecute(IBeaconSendingContext context)
...
+        }
+
         private void SendFinishedSessions(IBeaconSendingContext context)
         {
             // check if there's finished Sessions to be sent -> immediately send beacon(s) of finished Sessions
-            var finishedSession = context.GetNextFinishedSession();
-            while (finishedSession != null)
+            var finishedSessions = context.FinishedAndConfiguredSessions;
+
+            foreach (var finishedSession in finishedSessions)
```
Comment (y0chanan) [inline src/openkit-shared/Core/Communication/BeaconSendingCaptureOnState.cs]: when using 
var session in finishedSessions
the following code it is easier to read

File: src/openkit-shared/Core/Communication/SessionWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+        /// <summary>
+        /// Finish this session.
+        /// </summary>
+        internal void FinishSession()
+        {
+            sessionFinished = true;
+        }
+
+        /// Will be called each time a new session request was made for a session.
```
Comment (y0chanan) [inline src/openkit-shared/Core/Communication/SessionWrapper.cs]: like the other comments in this file this comment could be enclosed in ///<summary> and /// </summary>

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs [hunk 3]
```diff
@@ -152,16 +167,81 @@ public void TransitionToFlushStateIsPerformedOnShutdown()
...
+            target.Execute(context);
+
+            // verify for both new sessions a new session request has been made
+            httpClient.Received(2).SendNewSessionRequest();
+
+            // also verify that sessionOne got a new configuration
+            Assert.That(sessionOne.IsBeaconConfigurationSet, Is.True);
+            Assert.That(sessionOne.BeaconConfiguration.Multiplicity, Is.EqualTo(5));
+
+            // vor session two the number of requests was decremented
```
Comment (y0chanan) [inline tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs]: // for session

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingFlushSessionsStateTest.cs [hunk 4]
```diff
@@ -82,20 +85,71 @@ public void ShutdownStateIsTerminalState()
...
+        [Test]
+        public void ABeaconSendingFlushSessionsStateConfiguresAllUnconfiguredSessions()
+        {
+            // given
             var target = new BeaconSendingFlushSessionsState();
+
+            var sessionOne = new SessionWrapper(CreateValidSession("127.0.0.1"));
+            var sessionTwo = new SessionWrapper(CreateValidSession("127.0.0.2"));
+            var sessionThree = new SessionWrapper(CreateValidSession("127.0.0.2"));
+            sessionThree.End();
```
Comment (y0chanan) [inline tests/openkit-shared-tests/Core/Communication/BeaconSendingFlushSessionsStateTest.cs]: a comment that a session ended by the user is still unconfigured would ease the understanding why this statement is executed here

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs [hunk 5]
```diff
@@ -172,25 +252,24 @@ public void FinishedSessionsAreSent()
...
+            finishedSessions.Add(finishedSession);
             httpClient.SendBeaconRequest(Arg.Any<string>(), Arg.Any<byte[]>()).Returns((StatusResponse)null);
 
-
             //when calling execute
             target.Execute(context);
 
-            context.Received(1).GetNextFinishedSession();
-            context.Received(1).PushBackFinishedSession(finishedSession);
+            var tmp = context.Received(1).FinishedAndConfiguredSessions;
```
Comment (thomasgrassauer) [inline tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs:271]: `tmp` is never used

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs [hunk 6]
```diff
@@ -199,21 +278,22 @@ public void ABeaconSendingCaptureOnStateContinuesWithNextFinishedSessionIfSendin
...
-            context.GetNextFinishedSession().Returns(finishedEmptySession, finishedSession, null);
             httpClient.SendBeaconRequest(Arg.Any<string>(), Arg.Any<byte[]>()).Returns(new StatusResponse(string.Empty, 200));
-            
+
             //when calling execute
             target.Execute(context);
 
-            context.Received(3).GetNextFinishedSession();
-            context.Received(0).PushBackFinishedSession(finishedSession);
+            var tmp = context.Received(1).FinishedAndConfiguredSessions;
```
Comment (thomasgrassauer) [inline tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs:291]: `tmp` is never used

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs (hunk 5 above)
Comment (stefaneberl) [reply tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs:271]: Unfortunately an NSubstitute limitation http://nsubstitute.github.io/help/received-calls/
(see Checking calls to properties)

File: tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs (hunk 6 above)
Comment (stefaneberl) [reply tests/openkit-shared-tests/Core/Communication/BeaconSendingCaptureOnStateTest.cs:291]: Unfortunately an NSubstitute limitation http://nsubstitute.github.io/help/received-calls/
(see Checking calls to properties)

**Review by y0chanan — APPROVED** [review]

**Pull request merged**